File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.py
import heapq
def find(x):
    return 0
    

def solution(n, costs):
    parents = [i for i in range(n)]

    def find(x):
        nonlocal parents
        if x != parents[x]:
            return find(parents[x])
        else:
            return parents[x]
    def union(x,y):
        nonlocal parents

        x = find(x)
        y = find(y)
        if x< y:
            parents[y] = x
        else:
            parents[x] = y
    graph = []
    for c in costs:
        heapq.heappush(graph,[c[2],c[0],c[1]])
    graph.sort()
    ans = 0
    while graph:
        cur = heapq.heappop(graph)
        print(cur)
        print(parents)
        if find(cur[1]) != find(cur[2]):
            union(cur[1],cur[2])
            ans += cur[0]
        else:
            continue
    print(ans)
        
    answer = 0
    return ans

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.java
import java.util.*;
class Solution {
    static List<List<int[]>> relationList;
    public int solution(int n, int[][] costs) {
        int answer = 0;
        relationList = new ArrayList<>();
        for(int i=0;i<n;i++){
            relationList.add(new ArrayList<>());
        }
        for(int i=0;i<costs.length;i++){
            int n1 = costs[i][0];
            int n2 = costs[i][1];
            int cost = costs[i][2];
            relationList.get(n1).add(new int[]{cost,n2});
            relationList.get(n2).add(new int[]{cost,n1});
        }
        return prim(n);
    }
    public int prim(int n){
        PriorityQueue<String> pq = new PriorityQueue<>(new Comparator<String>(){
            @Override
            public int compare(String s1, String s2){
                String[] tmp1 = s1.split(" ");
                String[] tmp2 = s2.split(" ");
                int i1 = Integer.parseInt(tmp1[0]);
                int i2 = Integer.parseInt(tmp2[0]);
                return i1-i2;
            }
        });
        boolean[] visited = new boolean[n];
        pq.add("0 0");
        int ans = 0;
        while(!pq.isEmpty()){
            String cur = pq.poll();
            String[] tmp = cur.split(" ");
            int cost = Integer.parseInt(tmp[0]);
            int node = Integer.parseInt(tmp[1]);
            if(visited[node]){
                continue;
            }
            ans += cost;
            visited[node] = true;
            for(int[] list:relationList.get(node)){
                int nextC = list[0];
                int nextNode = list[1];
                pq.add(nextC+" "+nextNode);
            }
        }
        return ans;
    }
}